sympy/solvers/solveset.py

```diff
@@ -635,11 +634,12 @@ def _solveset(f, symbol, domain, _check=False):
             result = FiniteSet(*[s for s in result
                       if isinstance(s, RootOf)
                       or domain_check(fx, symbol, s)])
-
+    if _was_domain:
+        return Intersection(result, domain)
     return result
```

Comment:
> we can make every solve_* function do intersection independently

I advise against this to keep from computing intersections repeatedly. `solve_real_trig` should return a real solution and solveset should clip it in the right domain.